src/olympia/amo/middleware.py

```diff
@@ -279,38 +280,70 @@ def process_exception(self, request, exception):
-    will set HTTP_X_REQUEST_VIA_CDN to a secret value known to us so we can
-    verify the request did originate from the CDN.
-
-    If the request didn't come from the CDN and is a direct origin request, the
-    client IP will be in last position in HTTP_X_FORWARDED_FOR but in this case
-    nginx already sets REMOTE_ADDR so we don't need to use it.
+    Set REMOTE_ADDR from HTTP_X_FORWARDED_FOR if necessary.
+
+    The request flow is either:
+    Client -> CDN -> Load balancer -> WSGI proxy -> addons-server.
+    or
+    Client -> Load balancer -> WSGI proxy -> addons-server.
+
+    Currently:
+    - CDN is CloudFront
+    - Load Balancer is either a classic ELB (AWS) or GKE Ingress (GCP)
+    - WSGI proxy is nginx + uwsgi
+
+    CloudFront is set up to add a X-Request-Via-CDN header set to a secret
+    value known to us so we can verify the request did originate from the CDN.
+
+    Nginx converts X-Request-Via-CDN and X-Forwarded-For to
+    HTTP_X_REQUEST_VIA_CDN and HTTP_X_FORWARDED_FOR, respectively.
+
+    CloudFront always make make origin requests with a X-Forwarded-For header
+    set to "Client IP, CDN IP", so the client IP will be second to last for a
+    CDN request.
+
+    On top of that, on AWS, the classic ELB we're using does not make any
+    alterations to X-Forwarded-For but on GCP, GKE Ingress appends its own IP,
+    so the client IP would move from second to last to third to last in the CDN
+    case
+
+    If the request didn't come from the CDN and is a direct origin request, on
+    AWS the client IP will be in last position in HTTP_X_FORWARDED_FOR but in
+    this case nginx already sets REMOTE_ADDR so we don't need to use it. On GCP
+    though, that won't be true and we'll need to extract it from the second to
+    last position in HTTP_X_FORWARDED_FOR.
     """
 
     def is_request_from_cdn(self, request):
         return settings.SECRET_CDN_TOKEN and constant_time_compare(
             request.META.get('HTTP_X_REQUEST_VIA_CDN'), settings.SECRET_CDN_TOKEN
         )
 
+    def is_request_from_gcp_environement(self, request):
+        return os.environ.get('DEPLOY_PLATFORM') == 'gcp'
+
     def process_request(self, request):
+        position = 1
         x_forwarded_for_header = request.META.get('HTTP_X_FORWARDED_FOR')
-        if x_forwarded_for_header and self.is_request_from_cdn(request):
-            # We only have to split twice from the right to find what we're looking for.
-            try:
-                value = x_forwarded_for_header.rsplit(sep=',', maxsplit=2)[-2].strip()
-                if not value:
-                    raise IndexError
-                request.META['REMOTE_ADDR'] = value
-            except IndexError:
-                # Shouldn't happen, must be a misconfiguration, raise an error
-                # rather than potentially use/record incorrect IPs.
-                raise ImproperlyConfigured('Invalid HTTP_X_FORWARDED_FOR')
+        if x_forwarded_for_header:
```

Comment:
idea (that may not be an improvement, but putting it out there): split the header into `ips` and then `.pop()` off items off the end for each of the `if`s rather than iterating a numeric position variable

Reply:
I played around with this but it didn't really feel any better. I could do a `.split(',').pop(-position)` but that doesn't seem worth it.

Reply:
this is what I was thinking:
```python
def process_request(self, request):
        x_forwarded_for_header = request.META.get('HTTP_X_FORWARDED_FOR')
        if x_forwarded_for_header:
            ips = x_forwarded_for_header.split(',')
            try:
                if from_cdn := self.is_request_from_cdn(request):
                    ips.pop()
                if from_gcp := self.is_request_from_gcp_environement(request):
                    ips.pop()
                # If it's from the cdn or gcp we need to fix REMOTE_ADDR by
                # looking into HTTP_X_FORWARDED_FOR, otherwise we should ignore it.
                if from_cdn or from_gcp:
                    if remote_addr := ips.pop().strip():
                        request.META['REMOTE_ADDR'] = remote_addr
                    else:
                        raise IndexError
            except IndexError:
                # Shouldn't happen, must be a misconfiguration, raise an error
                # rather than potentially use/record incorrect IPs.
                raise ImproperlyConfigured('Invalid HTTP_X_FORWARDED_FOR')
```

Reply:
Not a fan - you have to keep track of how many `pop()` you did if you want to follow the code. I see the appeal of `from_gcp` and `from_cdn` though, but I'm not sure it's worth doing.

Reply:
The thinking was it was doing the reverse of what the CDN and/or GCP did but 🤷